File: SourceCode/C++/Sorting/Heapsort.cpp
#include "Maxheap.cpp"


/* *** ODSATag: Heapsort *** */
void heapsort(Comparable* A[], int n) {
  std::cout << "Getting started with array:" << std::endl;
  for (int j = 0; j<n; j++)
    std::cout << *A[j] << " ";
  std::cout << std::endl;
  MaxHeap H(A,n,n);
  std::cout << "Now, ready to unpack the heap" << std::endl;
  for (int i = 0; i < n; i++)
     H.removemax();
}
/* *** ODSAendTag: Heapsort *** */

// With KVPair

bool sorttest(int array[], int n, int threshold) {
  Comparable* A[n];
  int i;

  cout << "We came in with array:" << endl;
  for (i = 0; i<n; i++)
    std::cout << array[i] << " ";
  std::cout << std::endl;


  /* Sort an array of Ints */
  for (i = 0; i < n; ++i) {
    A[i] = new Int(array[i]);
  }

  cout << "We are going to call Heapsort with this array:" << endl;
  for (i = 0; i<n; i++)
    std::cout << A[i] << " ";
  std::cout << std::endl;

  //  for (i = 0; i < n; ++i) {
  //    cout << *A[i] << " ";
  //  }
  //  cout << std::endl;

  heapsort(A, n);

  if (!checkorder(A, n)) return false;

  for (i = 0; i < n; ++i) {
    delete A[i];
  }

  /* Sort an array of KVPairs */

  for (i = 0; i < n; ++i) {
    A[i] = new KVPair(array[i], &array[i]);
  }

  //heapsort(A, n);

  if (!checkorder(A, n)) return false;

  for (i = 0; i < n; ++i) {
    delete A[i];
  }

  delete[] array;

  return true;
}

#include "SortTest.cpp"

File: SourceCode/Java/Sorting/Heapsort.java
void sorttime(int[] B) {
    int i;
    long totaltime;
    int runs;
    double avgtime;

    if (!prod)
        System.out.println("Doing timings for an array of size " + B.length + " on the basis of " + numruns + " runs");

  // Timing test for standard implementation
    totaltime = 0;
    for (runs=0; runs<numruns; runs++) {
        for(i=0; i<B.length; i++) {
            A[i] = B[i];
        }
        time1 = System.nanoTime();
        heapsort(A);
        time2 = System.nanoTime();
        checkorder(A);
        totaltime += (time2-time1);
    }
    if (!prod)
        System.out.println("Total time is: " + totaltime + ", numruns is: " + numruns);
    avgtime = (((double)totaltime)/numruns) / 1000000.0;
    System.out.println("Standard Heapsort: Size " + A.length + ", Time: " + avgtime);
}


boolean sorttest(int[] B) {
  int i;
  for (i = 0; i < B.length; i++) {
    A[i] = B[i];
  }
  heapsort(A);
  if (!checkorder(A)) { return false; }
  return true;
}

/* *** ODSATag: Heapsort *** */
void heapsort(int[] A) {
  // The heap constructor invokes the buildheap method
  MaxHeap H = new MaxHeap(A, A.length, A.length);
  for (int i = 0; i < A.length; i++) {  // Now sort
    H.removeMax(); // Removemax places max at end of heap
  }
}
/* *** ODSAendTag: Heapsort *** */
